Enhance privacy of contactsmenu fixes #5107

LEDfan opened the pull request:
This filter outs all contacts which shouldn't be visible according to the rules in #5107.

LEDfan committed:
Privacy enhancements for contacts menu

 - Groups, which are excluded from sharing should not see local users at all
 - If sharing is restricted to users own groups, he should only see contacts from his groups:

Signed-off-by: Tobia De Koninck <tobia@ledfan.be>

LEDfan committed:
Fix tests

Signed-off-by: Tobia De Koninck <tobia@ledfan.be>

LEDfan committed:
Add tests

Signed-off-by: Tobia De Koninck <tobia@ledfan.be>

ChristophWurst commented on lib/private/Contacts/ContactsMenu/ContactsStore.php:
```diff
@@ -50,13 +70,62 @@ public function getContacts(IUser $user, $filter) {
+	 *  1. filter the current user
+	 *  2. if the `shareapi_exclude_groups` config option is enabled and the
+	 * current user is in an excluded group it will filter all local users.
+	 *  3. if the ``shareapi_only_share_with_group_members config option is
```
Should be
```
`shareapi_only_share_with_group_members`
```
I guess

ChristophWurst commented on lib/private/Contacts/ContactsMenu/ContactsStore.php:
```diff
@@ -50,13 +70,62 @@ public function getContacts(IUser $user, $filter) {
+	 * @return array the filtered contacts
+	 */
+	private function filterContacts(IUser $self, Array $entries) {
+		$excludedGroups = $this->config->getAppValue('core', 'shareapi_exclude_groups', 'no') === 'yes' ? true : false;
```
` ? true : false` is not necessary, the `===` check will already result in a boolean value.

ChristophWurst commented on lib/private/Contacts/ContactsMenu/ContactsStore.php:
```diff
@@ -50,13 +70,62 @@ public function getContacts(IUser $user, $filter) {
+		$excludedGroups = $this->config->getAppValue('core', 'shareapi_exclude_groups', 'no') === 'yes' ? true : false;
+
+		$skipLocal = false; // whether to filter out local users
+		$ownGroupsOnly = $this->config->getAppValue('core', 'shareapi_only_share_with_group_members', 'no') === 'yes' ? true : false; // whether to filter out all users which doesn't have the same group as the current user
```
same as above

ChristophWurst reviewed (changes requested):
Some early feedback. I haven't had a closer look yet, nor did I test this. Will do ASAP.
Thanks!

LEDfan committed:
Some code improvements

Signed-off-by: Tobia De Koninck <tobia@ledfan.be>

rotanid commented:
milestone nextcloud 13?? i would definitely call this serious privacy problem a regression that has to be fixed with the next maintenance release.

ChristophWurst commented on lib/private/Contacts/ContactsMenu/ContactsStore.php:
```diff
@@ -50,13 +70,63 @@ public function getContacts(IUser $user, $filter) {
+	 * enabled it will filter all users which doens't have a common group
+	 * with the current user.
+	 * @param IUser $self
+	 * @param $entries Entry[]
```
should be `@param Entry[] $entries`

ChristophWurst commented on lib/private/Contacts/ContactsMenu/ContactsStore.php:
```diff
@@ -50,13 +70,63 @@ public function getContacts(IUser $user, $filter) {
+	 * @param $entries Entry[]
+	 * @return array the filtered contacts
+	 */
+	private function filterContacts(IUser $self, Array $entries) {
```
`array` instead of `Array`

ChristophWurst commented on lib/private/Contacts/ContactsMenu/ContactsStore.php:
```diff
@@ -50,13 +70,63 @@ public function getContacts(IUser $user, $filter) {
+	 * with the current user.
+	 * @param IUser $self
+	 * @param $entries Entry[]
+	 * @return array the filtered contacts
```
`@return Entry[]`

ChristophWurst commented on lib/private/Contacts/ContactsMenu/ContactsStore.php:
```diff
@@ -50,13 +70,63 @@ public function getContacts(IUser $user, $filter) {
+	private function filterContacts(IUser $self, Array $entries) {
+		$excludedGroups = $this->config->getAppValue('core', 'shareapi_exclude_groups', 'no') === 'yes';
+
+		$skipLocal = false; // whether to filter out local users
```
move inline comments to a new line please

ChristophWurst commented on tests/lib/Contacts/ContactsMenu/ContactsStoreTest.php:
```diff
@@ -157,6 +175,135 @@ public function testGetContactsWithoutAvatarURI() {
+	}
+
+	public function testGetContactsOnlyIfInTheSameGroup() {
+		$this->config->expects($this->at(0)) ->method('getAppValue')
```
great to have this tested 👍

ChristophWurst reviewed (dismissed):
Tested and works as expected. Thanks a lot 👍 

I've left some comments regarding coding style.

LEDfan committed:
Improve code style

Signed-off-by: Tobia De Koninck <tobia@ledfan.be>

LEDfan commented:
@ChristophWurst I addressed your issues, thanks for testing.

@rotanid I do agree with you.
@karlitschek should this be backported to 12?

karlitschek commented:
i think this should be backported

LEDfan commented:
I just realized that the `ContactsStore` API probably should be public, so apps who interact with the contacts menu knows which contacts are available in the menu. E.g. in the Chat app we want to limit chatting to the users available in the Contacts menu (i.e. filtered by this PR)

An alternative way could be to move the filtering to the `ContactsManager` API.

What do you think?

LEDfan committed:
Fix issue when disabling the shareapi_only_share_with_group_members option + fix findOne

Signed-off-by: Tobia De Koninck <tobia@ledfan.be>

LEDfan commented:
@ChristophWurst I fixed a bug and also updated the findOne method because that wasn't filtered already.

LEDfan commented:
@MorrisJobke @LukasReschke @blizzz @rullzer @schiessle @nickvergessen @georgehrke Excuse me for pinging you all 😄 , but can someone please review this? There are many people asking for this.

Also please point out your opinion about https://github.com/nextcloud/server/pull/5585#issuecomment-320496777 and https://github.com/nextcloud/server/issues/5107#issuecomment-322807688

Thanks!

LukasReschke commented:
I'll review this and additionally revert https://github.com/nextcloud/server/commit/56a9084dd2f9a5726a74e1e789773558fade8b44 in here and add the handling here itself.

LukasReschke commented:
Rebased upon master.

LukasReschke committed:
Revert "Do not list system users in contacts menu if sharing autocompletion is disabled"

This reverts commit 56a9084dd2f9a5726a74e1e789773558fade8b44.

LukasReschke committed:
Add filter for `shareapi_allow_share_dialog_user_enumeration`

This adjusts the contacts menu to also support searching by email address which is relevant in scenarios where no UID is known such as LDAP, etc.

Furthermore, if `shareapi_allow_share_dialog_user_enumeration` is disabled only results are shown that match the full user ID or email address.

Signed-off-by: Lukas Reschke <lukas@statuscode.ch>

LukasReschke commented:
I added https://github.com/nextcloud/server/compare/5896176d69...705432ca6f, in particular:

- Searching for email addresses with `shareapi_allow_share_dialog_user_enumeration` is now possible again as it was in stable11: 

<img width="585" alt="screen shot 2017-09-15 at 16 02 36" src="https://user-images.githubusercontent.com/878997/30485955-55142afa-9a2f-11e7-820d-b8ca324b4229.png">
<img width="761" alt="screen shot 2017-09-15 at 16 02 42" src="https://user-images.githubusercontent.com/878997/30485956-55197744-9a2f-11e7-9e8a-6d9368715495.png">

- The contacts menu in the top right show contacts if a full UID or the full email has been searched for:

<img width="440" alt="screen shot 2017-09-15 at 16 03 41" src="https://user-images.githubusercontent.com/878997/30485993-75622d20-9a2f-11e7-8cf2-310f0a39e5e9.png">
<img width="399" alt="screen shot 2017-09-15 at 16 03 35" src="https://user-images.githubusercontent.com/878997/30485994-75b8eaac-9a2f-11e7-9e2a-8a9be026dd4f.png">

Note that the search is on "UID" and "EMAIL". So "FN" is not searched.

LukasReschke reviewed (approved):
Works 👍

danxuliu reviewed (approved):
CI complains because reverted commit was not signed, but as it is a revert commit it is not a big deal.

Tested and works; code looks good (but, please, next time do not mix code style changes and behaviour changes in the same commit!) 👍

LEDfan commented:
@LukasReschke thanks for taking care of this!

> Note that the search is on "UID" and "EMAIL". So "FN" is not searched.

Is there a reason for? E.g. on LDAP you as user would know the FN but not the username?

LukasReschke commented:
> Is there a reason for? E.g. on LDAP you as user would know the FN but not the UID?

Mainly because we have a business requirement requiring this behaviour: Filtering via UID / EMAIL should be possible but for whatever reason filtering to FN shouldn't be possible.

cc @oparoz Or would filtering for FN here also be acceptable? Your call 😉

LukasReschke commented:
Backport at https://github.com/nextcloud/server/pull/6554
